### PR TITLE
SWT Bot test cases for checking CEF property levels in CEFX

### LINKED ISSUE #112: Create SWT tests for CEF-X concept
Create SWT bot test cases for the tests from out test matrix.

### PR DESCRIPTION
**Pull request by franzTobiasDLR**
Checks that System, SubSystem and Equipments can only be added in correct levels of the tree.
It checks that by:
- UI command from the navigator
- Checkbox in the VirSatEditor

It furthermore checks that if levels are unchecked, corresponding artifacts are cleaned up properly.
 
Closes #112

### CODE REVIEW

**Comment by franzTobiasDLR** [comment]
Still checking what causes us to sometimes having to confirm changes, sometimes not...

**Comment by franzTobiasDLR** [comment]
> 
> 
> Still checking what causes us to sometimes having to confirm changes, sometimes not...

Create a ticket for the UX problem: https://github.com/virtualsatellite/VirtualSatellite4-CEF/issues/123

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/ACefSwtBotTestCase.java [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		super.before();
+		
+		conceptCef = ConceptXmiLoader.loadConceptFromPlugin(de.dlr.sc.virsat.model.extension.cef.Activator.PLUGIN_ID + "/concept/concept.xmi");
+		conceptCefX = ConceptXmiLoader.loadConceptFromPlugin(de.dlr.sc.virsat.model.extension.cefx.Activator.PLUGIN_ID + "/concept/concept.xmi");
+		
+	}
+	
+	@Override
+	protected void openCorePerspective() {
+		bot.menu("Window").menu("Perspective").menu("Open Perspective").menu("Other...").click();
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/ACefSwtBotTestCase.java:34]: Unnecessary code duplication

Comment (andriikovalov-dlr) [reply de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/ACefSwtBotTestCase.java:34]: Okay, this should be refactored in core

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 2]
```diff
@@ -0,0 +1,340 @@
...
+		
+		
+		SwtBotDebugHelper.logCodeLine();
+		bot.checkBox("System Mass Parameters").click();
+		SwtBotDebugHelper.logCodeLine();
+		waitForEditingDomainAndUiThread();
+		SwtBotDebugHelper.logCodeLine();
+		assertTrue(bot.checkBox("System Mass Parameters").isChecked());
+		SwtBotDebugHelper.logCodeLine();
+		final int EXPECTED_NUMBER_TREE_CHILDREN_SYSTEM_MATH_PARAMETERS = 4; // Documents folder, system parameters, system mass parameters, child SEI
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:98]: Let's remove THREE_CHILDREN from constant name, it doesn't match

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 3]
```diff
@@ -0,0 +1,340 @@
...
+		assertThat("There are documents, system parametes, system mass parameters and child items now", configurationTreeSystem.getItems(), arrayWithSize(EXPECTED_NUMBER_TREE_CHILDREN_SYSTEM_MATH_PARAMETERS));
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:99]: Are system parameters added automatically on adding mass parameters? Let's maybe explain it in a comment?

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 4]
```diff
@@ -0,0 +1,340 @@
...
+		SwtBotDebugHelper.logCodeLine();
+		bot.checkBox("Sub System Mass Parameters").click();
+		SwtBotDebugHelper.logCodeLine();
+		waitForEditingDomainAndUiThread();
+		bot.button("Save").click();
+		openEditor(elementConfigurationSubSystem);
+		SwtBotDebugHelper.logCodeLine();
+		assertFalse(bot.checkBox("Sub System Mass Parameters").isChecked());
+		SwtBotDebugHelper.logCodeLine();
+		final int EXPECTED_NUMBER_TREE_CHILDREN_SUB_SYSTEM = 2; // Documents folder, child SEI
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:147]: Name doesn't match the value

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 5]
```diff
@@ -0,0 +1,340 @@
...
+		SwtBotDebugHelper.logCodeLine();
+		openEditor(elementConfigurationEquipment);
+		SwtBotDebugHelper.logCodeLine();
+		bot.checkBox("Equipment Parameters").click();
+		SwtBotDebugHelper.logCodeLine();
+		waitForEditingDomainAndUiThread();
+		bot.button("Save").click();
+		
+		SwtBotDebugHelper.logCodeLine();
+		final int EXPECTED_NUMBER_TREE_CHILDREN_EQUIPMENT_PARAMETERS = 2; // Documents folder, system parameters
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:165]: Name doesn't match the value

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 6]
```diff
@@ -0,0 +1,340 @@
...
+		SwtBotDebugHelper.logCodeLine();
+		bot.checkBox("Equipment Parameters").click();
+		SwtBotDebugHelper.logCodeLine();
+		waitForEditingDomainAndUiThread();
+		bot.button("Save").click();
+		openEditor(elementConfigurationEquipment);
+		SwtBotDebugHelper.logCodeLine();
+		assertFalse(bot.checkBox("Equipment Parameters").isChecked());
+		SwtBotDebugHelper.logCodeLine();
+		final int EXPECTED_NUMBER_TREE_CHILDREN_EQUIPMENT = 1; // Documents folder
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:182]: Name totally doesn't match the value

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 7]
```diff
@@ -0,0 +1,340 @@
...
+		SwtBotDebugHelper.logCodeLine();
+		bot.checkBox("Equipment Mass Parameters").click();
+		SwtBotDebugHelper.logCodeLine();
+		waitForEditingDomainAndUiThread();
+		// Does not ask for save because equipment parameters are still there... So no hierarchy change happening
+		openEditor(elementConfigurationEquipment);
+		SwtBotDebugHelper.logCodeLine();
+		assertFalse(bot.checkBox("Equipment Mass Parameters").isChecked());
+		SwtBotDebugHelper.logCodeLine();
+		final int EXPECTED_NUMBER_TREE_CHILDREN_SYSTEM_PARAMETERS = 2; // Documents folder, system parameters
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:215]: Wrong name again

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java [hunk 8]
```diff
@@ -0,0 +1,340 @@
...
+	}
+	
+	@Test
+	public void testInitialEnabledLevels() {
+		
+		// Tests how CEF levels can be assigned before any level is assigned 
+		
+		// On configuration root level, the only the system CAs should be enabled
+		SwtBotDebugHelper.logCodeLine();
+		assertEnabled(configurationTreeSystem.contextMenu(conceptCefX.getDisplayName()).menu(getAddCommandName(SystemParameters.class)));
```
Comment (andriikovalov-dlr) [inline de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:227]: I suggest to refactor it to something like
`assertEnabled(addCommand(configurationTreeSystem, SystemParameters.class)`
to reduce visual noise

File: de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java (hunk 2 above)
Comment (franzTobiasDLR) [reply de.dlr.sc.virsat.cef.swtbot.test/src/de/dlr/sc/virsat/cef/swtbot/test/CefHierarchyLevelTest.java:98]: Well, its TREE children... Like ConfigurationTREE not the number ;)

**Comment by franzTobiasDLR** [comment]
Hm, now where I try to investigate the reason why it sometimes fails it's always succeding... We should collect information on these tests if it fails again later....

**Pull request merged**